### PR TITLE
Token service refactory

### PR DESCRIPTION
**Pull request by reasonerjt**
Due to the implementation in Docker client, seems it use the token for registry to access notary, I had use same token creator for both notary and registry, I will check the code of docker and discuss with them later.

### CODE REVIEW

File: src/ui/service/token/creator.go [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+	endpoint string
+}
+
+func (e endpointParser) parse(s string) (*image, error) {
+	repo := strings.SplitN(s, "/", 2)
+	if len(repo) < 2 {
+		return nil, fmt.Errorf("Unable to parse image from string: %s", s)
+	}
+	//Workaround, need to use endpoint Parser to handle both cases.
+	if strings.ContainsRune(repo[0], '.') {
```
Comment (ywk253100) [inline src/ui/service/token/creator.go]: It seems that '.' is a valid character in repo name.

File: src/ui/service/token/creator.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+	if len(scopeParm) > 0 {
+		scopes = strings.Split(r.URL.Query()["scope"][0], " ")
+	}
+	log.Debugf("scopes: %v", scopes)
+	for _, v := range g.validators {
+		user, err = v.validate(r)
+		if user != nil {
+			break
+		}
+		if err != nil {
```
Comment (ywk253100) [inline src/ui/service/token/creator.go]: Why put the error handling logic after the judgement statement?

File: src/ui/service/token/creator.go [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+		if len(scopes) == 0 {
+			return nil, &unauthorizedError{}
+		}
+		user = &userInfo{}
+	}
+	access := GetResourceActions(scopes)
+	for _, a := range access {
+		f, ok := g.filterMap[a.Type]
+		if !ok {
+			log.Warningf("No filter found for access type: %s, skip.", a.Type)
```
Comment (ywk253100) [inline src/ui/service/token/creator.go]: Does there need a continue?

Comment (reasonerjt) [inline src/ui/service/token/creator.go]: yes,thanks

File: src/ui/service/token/creator.go (hunk 1 above)
Comment (reasonerjt) [inline src/ui/service/token/creator.go]: This is the first chunk, the parser makes best effort to guess if it starts with a host name.  I think docker has the same logic?

Comment (reasonerjt) [inline src/ui/service/token/creator.go]: I'll fix it so that if endpoints mismatch it falls back to basic parser.

**Pull request merged**